File: 2022/4/solution.py
import re
import os

file1 = open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'input.txt'),'r')
lines = file1.readlines()
n = 0

for line in lines:
    nums = list(map(int, re.findall('\d+', line)))
    if (nums[0] >= nums[2] and nums[1] <= nums[3]) or (nums[0] <= nums[2] and nums[1] >= nums[3]) :
        n += 1

print(n)

n2 = 0

for line in lines:
    nums = list(map(int, re.findall('\d+', line)))
    if nums[0] <= nums[2] <= nums[1] or nums[0] <= nums[3] <= nums[1] or nums[2] <= nums[0] <= nums[3] or nums[2] <= nums[1] <= nums[3]:
        n2 += 1

print(n2)

File: 2022/4/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/weitecklee/adventofcode/utils"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func parseInput(data []string) [][2][2]int {
	puzzleInput := make([][2][2]int, len(data))
	for i, line := range data {
		nums := utils.ExtractInts(line)
		puzzleInput[i] = [2][2]int{{nums[0], nums[1]}, {nums[2], nums[3]}}
	}
	return puzzleInput
}

func fullyContained(pairs [2][2]int) bool {
	a, b := pairs[0][0], pairs[0][1]
	c, d := pairs[1][0], pairs[1][1]
	return (c <= a && c <= b && d >= a && d >= b) || (a <= c && a <= d && b >= c && b >= d)
}

func hasOverlap(pairs [2][2]int) bool {
	a, b := pairs[0][0], pairs[0][1]
	c, d := pairs[1][0], pairs[1][1]
	return (a >= c && a <= d) || (b >= c && b <= d) || (c >= a && c <= b) || (d >= a && d <= b)
}

func part1(puzzleInput [][2][2]int) int {
	res := 0
	for _, pairs := range puzzleInput {
		if fullyContained(pairs) {
			res++
		}
	}
	return res
}

func part2(puzzleInput [][2][2]int) int {
	res := 0
	for _, pairs := range puzzleInput {
		if hasOverlap(pairs) {
			res++
		}
	}
	return res
}
